kentik_api_library/kentik_api/synthetics/api_transport_grpc.py

```diff
@@ -0,0 +1,880 @@
+        )
+        response = self._data.GetHealthForTests(request=get_health_req, metadata=self._credentials, target=self._url)
+        return pb_to_health_tests(response.health)
+
+    def get_trace_for_test(self, **kwargs) -> GetTraceForTestResponse:
+        id = str(kwargs["id"])
+        agents = [str(id) for id in kwargs["agent_ids"]]
+        ips = [str(ip) for ip in kwargs["target_ips"]]
+        start = Timestamp(seconds=int(kwargs["start_time"].timestamp()))
+        end = Timestamp(seconds=int(kwargs["end_time"].timestamp()))
+        get_trace_req = GetTraceForTestRequest(
+            id=id,
+            start_time=start,
+            end_time=end,
+            agent_ids=agents,
+            target_ips=ips,
+        )
+        response = self._data.GetTraceForTest(request=get_trace_req, metadata=self._credentials, target=self._url)
+        return pb_to_trace_response(response)
+
+
+def populate_test_from_pb(v: pbTest, out: SynTest) -> None:
+    out.name = v.name
+    out.type = TestType(v.type)
+    out.status = PB_TO_TEST_STATUS[v.status]
+    out.deviceId = ID(v.device_id)
+    out._id = ID(v.id)
+    out._cdate = pb_to_datetime_iso(v.cdate)
+    out._edate = pb_to_datetime_iso(v.edate)
+
+    settings = SynTestSettings()
+    pupulate_settings_from_pb(v.settings, settings)
+    out.settings = settings
+
+
+# disable no-member linting as gRPC structures are wrongly recognized as to have missing attributes
+# pylint: disable=no-member
+def pb_from_test(v: SynTest) -> pbTest:
+    out = pbTest()
+    out.name = v.name
+    out.device_id = str(v.deviceId)
+    out.id = str(v.id)
+    out.type = v.type.value
+    out.status = reverse_map(PB_TO_TEST_STATUS, v.status)
+    out.settings.CopyFrom(pb_from_settings(v.settings))
+    return out
+
+
+def pupulate_ping_trace_test_from_pb(v: pbTest, out: PingTraceTest) -> None:
+    populate_test_from_pb(v, out)
+    populate_ping_test_settings_from_pb(v.settings, out.settings)
+
+
+def populate_mesh_test_from_pb(v: pbTest, out: MeshTest) -> None:
+    pupulate_ping_trace_test_from_pb(v, out)
+    # nothing more to populate
+
+
+def pupulate_settings_from_pb(v: pbTestSettings, out: SynTestSettings) -> None:
+    out.agentIds = [ID(id) for id in v.agent_ids]
+    out.tasks = v.tasks
+    out.healthSettings = pb_to_health_settings(v.health_settings)
+    out.monitoringSettings = pb_to_monitoring_settings(v.monitoring_settings)
+    out.port = v.port
+    out.period = v.period
+    out.count = v.count
+    out.expiry = v.expiry
+    out.limit = v.limit
+    out.protocol = Protocol(v.protocol)
+    out.family = PB_TO_IP_FAMILY[v.family]
+    out.rollupLevel = v.rollup_level
+    out.servers = v.servers
+
+
+# disable no-member linting as gRPC structures are wrongly recognized as to have missing attributes
+# pylint: disable=no-member
+def pb_from_settings(v: SynTestSettings) -> pbTestSettings:
+    out = pbTestSettings()
+
+    # TODO: add missing fields to SynTestSettings
```

Comment:
Which "missing fields".

Settings for `ping` and `trace` tasks are in `PingTraceTestSettings` (via `PingTask` and `TraceTask`). Note that there are tests that do not have any ping and trace tasks (e.g `dns`), so this cannot be in `SynTestSettings`.

This also nicely illustrates the serialization of tests settings to protobufs must be test type specific, hence `to_pb` instance method in `SynTest` and derived classes. Using the functional approach, you'll have to have `pb_from_<test_type>_settings` functions (there are actually fewer variations than there are test types, but still you need more than one).

Reply:
Will update while migrating to API 2022